File: Python/Prototype.py
"""
"
" date: 2016/8/29
" author: Weian Cheng
" describe: practice Prototype Pattern.
"           the key point is clone(), I used copy() to implement this patter.
"
"""
#!/usr/bin/env python3

from abc import abstractmethod
import copy


class Matrix(object):
    @abstractmethod
    def show(self):
        pass

    @abstractmethod
    def clone(self):
        pass


class Vector3(Matrix):
    def __init__(self, x=0, y=0, z=0):
        self._x = x
        self._y = y
        self._z = z

    @property
    def x(self):
        return self._x

    @property
    def y(self):
        return self._y

    @property
    def z(self):
        return self._z

    @x.setter
    def x(self, value):
        self._x = value

    @y.setter
    def y(self, value):
        self._y = value

    @z.setter
    def z(self, value):
        self._z = value

    def show(self):
        print("Vector information:")
        print("X: " + str(self._x))
        print("Y: " + str(self._y))
        print("Z: " + str(self._z))

    def clone(self):
        return copy.copy(self)  # only clone this object
        # clone this object, and clone another object.
        # return copy.deepcopy(self)


def main():
    v = Vector3()
    v.x = 10
    v.y = 20
    v.z = 30
    print("Vector1: ")
    v.show()

    print("")

    v2 = Vector3(50, 50, 50)
    print("Vector2: ")
    v2.show()

    v2 = v.clone()
    print("\nVector3: ")
    v2.show()

if __name__ == "__main__":
    main()

File: Kotlin/Prototype.kt
interface Profile {
    fun show()
    fun clone(): Profile
}

class Vector3(x: Float, y: Float, z: Float): Profile {
    private val x = x
    private val y = y
    private val z = z
    override public fun show() {
        println("x: " + x + ", y: " + y + ", z: " + z)
    }
    override public fun clone(): Profile {
        return Vector3(x, y, z)
    }
}

class Rotation(w: Float, x: Float, y: Float, z: Float): Profile {
    private val w = w
    private val x = x
    private val y = y
    private val z = z
    override public fun show() {
        println("w: " + w + ", x: " + x + ", y: " + y + ", y: " + y + ", z: " + z)
    }
    override public fun clone(): Profile {
        return Rotation(w, x, y, z)
    }
}

fun main() {
    println("This is Prototype Pattern")
    val v1 = Vector3(1.1f, 2.2f, 3.3f)
    v1.show()
    val v2 = v1.clone()
    v2.show()
    val r1 = Rotation(4.4f, 5.5f, 6.6f, 7.7f)
    r1.show()
    val r2 = r1.clone()
    r2.show()
}
